backend/routes/indexer.go

```diff
@@ -58,86 +58,145 @@ func InitIndexerRoutes() {
-	colorHex := reqBody["data"].(map[string]interface{})["batch"].([]interface{})[0].(map[string]interface{})["events"].([]interface{})[0].(map[string]interface{})["event"].(map[string]interface{})["data"].([]interface{})[0]
+	// Extract necessary fields from the request body
+	data, ok := reqBody["data"].(map[string]interface{})
+	if !ok {
+		handleError(w, "Invalid request format (missing 'data' field)", http.StatusBadRequest)
+		return
+	}
+
+	batch, ok := data["batch"].([]interface{})
+	if !ok || len(batch) == 0 {
+		handleError(w, "Invalid request format (missing or empty 'batch' array)", http.StatusBadRequest)
+		return
+	}
+
+	event, ok := batch[0].(map[string]interface{})["events"].([]interface{})
+	if !ok || len(event) == 0 {
+		handleError(w, "Invalid request format (missing or empty 'events' array)", http.StatusBadRequest)
+		return
+	}
+
+	eventData, ok := event[0].(map[string]interface{})
+	if !ok {
+		handleError(w, "Invalid event format (missing 'event' object)", http.StatusBadRequest)
+		return
+	}
+
+	keys, ok := eventData["keys"].([]interface{})
+	if !ok || len(keys) < 4 {
+		handleError(w, "Invalid event format (missing or incomplete 'keys' array)", http.StatusBadRequest)
+		return
+	}
 
-	// Convert hex to int
-	position, err := strconv.ParseInt(posHex.(string), 0, 64)
+	// Extract and validate address, position, day index, and color
+	address := getStringValue(keys, 1)
+	posHex := getStringValue(keys, 2)
+	dayIdxHex := getStringValue(keys, 3)
+	colorHex := getStringValue(eventData["data"], 0)
+
+	position, err := strconv.ParseInt(posHex, 0, 64)
 	if err != nil {
-		fmt.Println("Error converting position hex to int: ", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		handleError(w, "Error converting position hex to int", http.StatusBadRequest)
 		return
 	}
-	dayIdx, err := strconv.ParseInt(dayIdxHex.(string), 0, 64)
+
+	dayIdx, err := strconv.ParseInt(dayIdxHex, 0, 64)
 	if err != nil {
-		fmt.Println("Error converting day index hex to int: ", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		handleError(w, "Error converting day index hex to int", http.StatusBadRequest)
 		return
 	}
-	color, err := strconv.ParseInt(colorHex.(string), 0, 64)
+
+	color, err := strconv.ParseInt(colorHex, 0, 64)
 	if err != nil {
-		fmt.Println("Error converting color hex to int: ", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		handleError(w, "Error converting color hex to int", http.StatusBadRequest)
+		return
+	}
+
+	// Validate position and color
+	if position < 0 || position >= int64(core.ArtPeaceBackend.CanvasConfig.Canvas.Width*core.ArtPeaceBackend.CanvasConfig.Canvas.Height) {
+		handleError(w, "Invalid position value", http.StatusBadRequest)
+		return
+	}
+
+	if color < 0 || color >= (1 << core.ArtPeaceBackend.CanvasConfig.ColorsBitWidth) {
```

Comment:
Checking against `core.ArtPeaceBackend.CanvasConfig.Colors` length would be better here.

Reply:
core.ArtPeaceBackend.CanvasConfig.Colors.length, Noted.